Development/nmos/ocsp_behaviour.cpp

```diff
@@ -0,0 +1,402 @@
+
+                    slog::log<slog::severities::info>(gate, SLOG_FLF) << "cache the OCSP response";
+
+                    nmos::with_write_lock(*ocsp_settings.mutex, [&] { ocsp_settings.ocsp_response = ocsp_resp; });
+
+                    model.notify();
+
+                    return true;
+                });
+            }).then([&](pplx::task<void> finally)
+            {
+                auto lock = model.write_lock(); // in order to update local state
+
+                try
+                {
+                    finally.get();
+                }
+                catch (const web::http::http_exception& e)
+                {
+                    slog::log<slog::severities::error>(gate, SLOG_FLF) << "Certificate status request HTTP error: " << e.what() << " [" << e.error_code() << "]";
+                }
+                catch (const ocsp_service_exception&)
+                {
+                    slog::log<slog::severities::error>(gate, SLOG_FLF) << "Certificate status request error";
+                }
+
+                // reaching here, there must be something has gone wrong with the OCSP API
+                // let's select the next available OCSP API
+                state.ocsp_service_error = true;
+
+                model.notify();
+            });
+        }
+
+        int half_certificate_expiry_from_now(const std::vector<utility::string_t>& certificate_chains_data, slog::base_gate& gate)
+        {
+            int expiry_time = -1;
+            try
+            {
+                // get the shorter expiry time from all the certificates
+                for (const auto& cert_data : certificate_chains_data)
+                {
+                    const auto expiry_time_ = ssl::experimental::certificate_expiry_from_now(utility::us2s(cert_data), 0.5);
+                    expiry_time = expiry_time < 0 ? expiry_time_ : std::min(expiry_time, expiry_time_);
+                }
+            }
+            catch (const ssl::experimental::ssl_exception& e)
+            {
+                slog::log<slog::severities::error>(gate, SLOG_FLF) << "Get certificate expiry time with SSL error: " << e.what();
+            }
+            catch (const std::exception& e)
+            {
+                slog::log<slog::severities::error>(gate, SLOG_FLF) << "Get certificate expiry time with unexpected exception: " << e.what();
+            }
+            catch (...)
+            {
+                slog::log<slog::severities::severe>(gate, SLOG_FLF) << "Get certificate expiry time with nexpected unknown exception";
+            }
+            return (expiry_time < 0 ? 0 : expiry_time);
+        }
+
+        // construct a list of OCSP URIs from a list of server certificate chains
+        std::vector<web::uri> get_ocsp_uris(const std::vector<utility::string_t>& certificate_chains_data, slog::base_gate& gate)
+        {
+            std::vector<web::uri> ocsp_uris;
+
+            try
```

Comment:
Does this `try`...`catch` block belong in this function (and `make_ocsp_request`) or would it be better closer to 'the edge'?